src/binary/ibc-setup/commands/balances.ts

```diff
@@ -0,0 +1,79 @@
+import os from 'os';
+import path from 'path';
+
+import { stringToPath } from '@cosmjs/crypto';
+import { GasPrice } from '@cosmjs/launchpad';
+import { DirectSecp256k1HdWallet } from '@cosmjs/proto-signing';
+import { Coin } from '@cosmjs/proto-signing/build/codec/cosmos/base/v1beta1/coin';
+
+import { IbcClient } from '../../../lib/ibcclient';
+import { registryFile } from '../../constants';
+import { loadAndValidateApp } from '../utils/load-and-validate-app';
+import { loadAndValidateRegistry } from '../utils/load-and-validate-registry';
+import { resolveHomeOption } from '../utils/options/shared/resolve-home-option';
+import { resolveKeyFileOption } from '../utils/options/shared/resolve-key-file-option';
+import { resolveMnemonicOption } from '../utils/options/shared/resolve-mnemonic-option';
+
+import { Flags, getAddresses, Options } from './keys-list';
+
+export async function balances(flags: Flags) {
+  const home = resolveHomeOption({ homeFlag: flags.home });
+  const app = loadAndValidateApp(home);
+  const keyFile = resolveKeyFileOption({ keyFileFlag: flags.keyFile, app });
+  const mnemonic = await resolveMnemonicOption({
+    interactiveFlag: flags.interactive,
+    mnemonicFlag: flags.mnemonic,
+    keyFile: keyFile,
+    app,
+  });
+
+  const options: Options = {
+    home,
+    mnemonic,
+  };
+
+  await run(options);
+}
+
+export async function run(options: Options) {
+  const registryFilePath = path.join(options.home, registryFile);
+  const registry = loadAndValidateRegistry(registryFilePath);
+
+  const addresses = await getAddresses(registry.chains, options.mnemonic);
+
+  const balances = (
+    await Promise.all(
+      addresses.map<Promise<[string, Coin]>>(async ([chain, data, address]) => {
+        const signer = await DirectSecp256k1HdWallet.fromMnemonic(
```

Comment:
These lines (47-63) make a IbcClient from the chain data. This is a great function to pull out.

You don't need to even provide address here, you can get that from the signer (like in the getAddresses function)

Reply:
If you do that, you could then do something like:

```ts
registry.chains.map(([chain, data]) => {
  const {client, address, feeToken} = await buildIbcClient(data);
  const coin = await client.query.bank.unverified.balance(address, feeToken);
  return [chain, coin];
})....
```

Reply:
But yeah, that can also be in another PR (to pull it out).

Given one chain data to build the IbcClient will be needed for most of the other functions. You can merge and leave it to Will to refactor this for his command